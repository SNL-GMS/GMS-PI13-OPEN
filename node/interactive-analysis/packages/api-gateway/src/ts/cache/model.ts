import Immutable from 'immutable';
import { Association, TimeRange } from '../common/model';
import { Event } from '../event/model-and-schema/model';
import { QcMask } from '../qc-mask/model';
import { SignalDetection } from '../signal-detection/model';
import { UserProfile } from '../user-profile/model';

/**
 * Specifies the type (method) for committing to the Global Cache.
 */
export type CommitFunc<T> = (items: T[]) => void;

/**
 * User Cache Processor interface.
 * Defines all of the common method definitions required
 * for a user cache item.
 */
// tslint:disable-next-line: interface-name
export interface IUserCacheItem<T> {
  has(id: string): boolean;
  get(id: string): T;
  getAll(): T[];
  set(item: T): void;
  setAll(items: T[]): void;
  remove(item: T): void;
  removeAll(item: T[]): void;
  commitAll(): void;
  commitWithIds(ids: string[]): void;
  updateFromGlobalCache(cachedEntries: T[], overwrite: boolean): void;
}

/**
 * User Cache interface.
 * Defines all of the common method definitions required for a user cache.
 */
// tslint:disable-next-line: interface-name
export interface IUserCache {
  getTimeRange(): TimeRange;
  setTimeRange(timeRange: TimeRange): void;
  getUserProfile(): UserProfile;
  setUserProfile(userProfile: UserProfile): void;
  getOpenEventId(): string;
  setOpenEventId(id: string): void;
  getEventById(id: string): Event | undefined;
  getEvents(): Event[];
  getInvalidEventIds(): string[];
  setEvent(userAction: UserActionDescription, event: Event): void;
  setEvents(userAction: UserActionDescription, events: Event[]): void;
  updateEventsFromGlobalCache(events: Event[], overwrite: boolean): void;
  commitAllEvents(): void;
  commitEventsWithIds(ids: string[]): void;
  getSignalDetectionById(id: string): SignalDetection | undefined;
  getSignalDetections(): SignalDetection[];
  getInvalidSignalDetectionIds(): string[];
  setSignalDetection(userAction: UserActionDescription, sd: SignalDetection): void;
  setSignalDetections(userAction: UserActionDescription, sds: SignalDetection[]): void;
  updateSignalDetectionsFromGlobalCache(sds: SignalDetection[], overwrite: boolean): void;
  commitAllSignalDetections(): void;
  commitSignalDetectionsWithIds(ids: string[]): void;
  getAssociations(): Association[];
  getSignalDetectionAndEventBySdId(id): { signalDetection: SignalDetection; event: Event };
  setEventsAndSignalDetections(
    userAction: UserActionDescription,
    events: Event[],
    sds: SignalDetection[]
  ): void;
  getRedoPriorityOrder(id: string): number | undefined;
  getHistory(): History[];
  undoHistory(numberOfItems: number): UserCacheData;
  redoHistory(numberOfItems: number): UserCacheData;
  undoHistoryById(id: string): UserCacheData;
  redoHistoryById(id: string): UserCacheData;
  getEventHistory(id: string): History[];
  undoEventHistory(numberOfItems: number): UserCacheData;
  redoEventHistory(numberOfItems: number): UserCacheData;
  undoEventHistoryById(id: string): UserCacheData;
  redoEventHistoryById(id: string): UserCacheData;
  clearHistory(): void;
}

/**
 * User context for apollo and user sessions
 */
export interface UserContext {
  readonly sessionId: string;
  readonly userName: string;
  readonly userCache: IUserCache;
  readonly userRole: string;
}

/**
 * Represents the user cache data from the user cache.
 */
export interface UserCacheData {
  readonly events: Event[];
  readonly signalDetections: SignalDetection[];
}

/**
 * Description of a user action preformed
 */
export enum UserActionDescription {
  UNKNOWN = 'Unknown',
  CHANGE_SIGNAL_DETECTION_ASSOCIATIONS_ASSOCIATE = 'Associate',
  CHANGE_SIGNAL_DETECTION_ASSOCIATIONS_ASSOCIATE_MULTIPLE = 'Associate multiple detections',
  CHANGE_SIGNAL_DETECTION_ASSOCIATIONS_UNASSOCIATE = 'Unassociate',
  CHANGE_SIGNAL_DETECTION_ASSOCIATIONS_UNASSOCIATE_MULTIPLE = 'Unassociate multiple detections',

  // signal detection actions
  CREATE_DETECTION = 'Create detection',
  REJECT_DETECTION = 'Reject detection',
  REJECT_MULTIPLE_DETECTIONS = 'Reject multiple detections',
  UPDATE_DETECTION_RE_TIME = 'Time',
  UPDATE_DETECTION_RE_PHASE = 'Phase',
  UPDATE_MULTIPLE_DETECTIONS_RE_PHASE = 'Phase multiple detections',
  UPDATE_DETECTION_AMPLITUDE = 'Update detection amplitude',
  UPDATE_DETECTION_REVIEW_AMPLITUDE = 'Update detection review amplitude',
  UPDATE_DETECTION = 'Update detection',
  COMPUTE_FK = 'Fk',
  COMPUTE_MULTIPLE_FK = 'Fk multiple detections',

  // event actions
  CREATE_EVENT = 'Create event',
  UPDATE_EVENT_LOCATE = 'Locate event',
  UPDATE_EVENT_FROM_SIGNAL_DETECTION_CHANGE = 'Update event from signal detection change',
  UPDATE_EVENT_FEATURE_PREDICTIONS = 'Update event feature predictions',
  UPDATE_EVENT_STATUS_OPEN_FOR_REFINEMENT = 'Event opened for refinement',
  UPDATE_EVENT_MARK_COMPLETE = 'Event marked complete',
  UPDATE_EVENT_PREFERRED_HYP = 'Update event preferred hypothesis',
  UPDATE_EVENT_MAGNITUDE = 'Change Magnitude defining settings',
  SAVE_EVENT = 'Save event'
}

/**
 * Description of a user action performed that cannot be undone and will not be in the history
 */
export const NON_UNDO_ACTIONS: UserActionDescription[] = [
  UserActionDescription.UPDATE_EVENT_FEATURE_PREDICTIONS,
  UserActionDescription.UPDATE_EVENT_STATUS_OPEN_FOR_REFINEMENT,
  UserActionDescription.UPDATE_EVENT_MARK_COMPLETE,
  UserActionDescription.SAVE_EVENT,
  UserActionDescription.UPDATE_DETECTION_REVIEW_AMPLITUDE
];

/**
 * Defines the interface of User Action.
 */
export interface UserAction {
  /**
   * The user action description
   */
  readonly description: UserActionDescription;

  /**
   * Returns the toString value of a user action.
   *
   * @returns the string representation of the action
   */
  toString(): string;
}

/**
 * Defines the hypothesis type for a hypothesis change.
 */
export enum HypothesisType {
  EventHypothesis = 'EventHypothesis',
  SignalDetectionHypothesis = 'SignalDetectionHypothesis'
}

/**
 * Represents hypothesis information for a change.
 * Defines a mapping from a hypothesis id to the main
 * object id.
 */
export interface HypothesisChangeInformation {
  readonly id: string;
  readonly hypothesisId: string;
  readonly type: HypothesisType;
  readonly parentId: string;
  readonly userAction: UserAction;
}

/**
 * Represents a history change.
 */
export interface HistoryChange {
  readonly id: string;
  readonly active: boolean;
  readonly eventId: string;
  readonly conflictCreated: boolean;
  readonly hypothesisChangeInformation: HypothesisChangeInformation;
}

/**
 * Represents the history.
 */
export interface History {
  readonly id: string;
  readonly description: UserActionDescription;
  readonly changes: Immutable.List<HistoryChange>;
}

/**
 * Represents data that has been created or updated and needs to
 * be sent to the UI Clients
 *
 * Currently only include events and sds
 */
export interface DataPayload {
  readonly events: Event[];
  readonly sds: SignalDetection[];
  readonly qcMasks: QcMask[];
}

/**
 * Invalid data; typically data that has been deleted.
 */
export interface InvalidData {
  readonly eventIds: string[];
  readonly signalDetectionIds: string[];
}
