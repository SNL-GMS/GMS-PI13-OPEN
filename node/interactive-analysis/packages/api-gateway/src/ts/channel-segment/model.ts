import { PhaseType } from '../common/model';
import { FkPowerSpectra, FkPowerSpectraOSD } from '../fk/model';
import { ProcessingChannel } from '../station/processing-station/model';
import { OSDWaveform, Waveform } from '../waveform/model';

/**
 * Represents a channel time segment comprised of one or more waveforms for a (derived or raw).
 * Channel Segment definitions shared across gateway data APIs
 */
export interface ChannelSegment<T extends TimeSeries> {
  id: string;
  channel: ProcessingChannel;
  name: string;
  type: ChannelSegmentType;
  timeseriesType: TimeSeriesType;
  startTime: number;
  endTime: number;
  timeseries: T[];
  requiresSave: boolean;
}

/**
 * OSD channel segment as a generic of OSD time series
 */
export interface OSDChannelSegment<T extends OSDTimeSeries> {
  id: string;
  channel: ProcessingChannel;
  name: string;
  type: ChannelSegmentType;
  timeseriesType: TimeSeriesType;
  startTime: string;
  endTime: string;
  timeseries: T[];
}

/**
 * Channel segment input
 */
export interface ChannelSegmentInput {
  startTime: string;
  endTime: string;
  channelNames: string[];
  withWaveforms: boolean;
}

/**
 * Channel segment by id input
 */
export interface ChannelSegmentByIdInput {
  channelSegmentIds: string[];
  withTimeseries: boolean;
}

/**
 * Represents a generic time series data set output from a channel
 * (derived or raw) within a specified time range.
 */
export interface TimeSeries {
  startTime: number;
  sampleRate: number;
  sampleCount: number;
  type: TimeSeriesType;
}

/**
 * Time series OSD representation
 */
export interface OSDTimeSeries {
  startTime: string;
  sampleRate: number;
  sampleCount: number;
  type: TimeSeriesType;
}

/**
 * Channel Segment type
 */
export enum ChannelSegmentType {
  ACQUIRED = 'ACQUIRED',
  RAW = 'RAW',
  DETECTION_BEAM = 'DETECTION_BEAM',
  FK_BEAM = 'FK_BEAM',
  FK_SPECTRA = 'FK_SPECTRA',
  FILTER = 'FILTER'
}

/**
 * Time series type
 */
export enum TimeSeriesType {
  WAVEFORM = 'WAVEFORM',
  FK_SPECTRA = 'FK_SPECTRA',
  DETECTION_FEATURE_MAP = 'DETECTION_FEATURE_MAP'
}

/**
 * Beam Input Definitions for Compute Beam Streaming Call
 */
export interface BeamFormingInput {
  beamDefinition: BeamDefinition;
  // processingContext: ProcessingContext;
  // outputChannelId: string;
  waveforms: OSDChannelSegment<OSDTimeSeries>[];
}

/**
 * Beam Definition used by BeamFormingInput
 */
export interface BeamDefinition {
  phaseType: PhaseType;
  azimuth: number;
  slowness: number;
  coherent: boolean;
  snappedSampling: boolean;
  twoDimensional: boolean;
  nominalWaveformSampleRate: number;
  waveformSampleRateTolerance: number;
  // beamPoint: Location;
  // OSD "map" is not really a map so type as any
  // relativePositionsByChannelId: any;
  minimumWaveformsForBeam: number;
}

/**
 * Checks if FK spectra channel segment
 * @param object Channel Segment
 * @returns boolean
 */
// tslint:disable-next-line:max-line-length
export function isFkSpectraChannelSegment(
  object: ChannelSegment<TimeSeries>
): object is ChannelSegment<FkPowerSpectra> {
  return object.timeseriesType === TimeSeriesType.FK_SPECTRA;
}

/**
 * Checks id FK spectra channel segment OSD representation
 * @param object OSDChannelSegment
 * @returns boolean
 */
// tslint:disable-next-line:max-line-length
export function isFkSpectraChannelSegmentOSD(
  object: OSDChannelSegment<OSDTimeSeries>
): object is OSDChannelSegment<FkPowerSpectraOSD> {
  return object.timeseriesType === TimeSeriesType.FK_SPECTRA;
}

/**
 * Checks if FK Spectra time series
 * @param object Time Series
 * @param type Time Series Type
 * @returns boolean
 */
export function isFkSpectraTimeSeries(
  object: TimeSeries,
  type: TimeSeriesType
): object is FkPowerSpectra {
  return type === TimeSeriesType.FK_SPECTRA;
}

/**
 * Checks if FK Spectra time series
 * @param object Time Series
 * @param type Time Series Type
 * @returns boolean
 */
export function isWaveformTimeSeries(object: TimeSeries, type: TimeSeriesType): object is Waveform {
  return type === TimeSeriesType.WAVEFORM;
}

/**
 * Checks if waveform channel segment
 * @param object Channel segment
 * @returns boolean
 */
export function isWaveformChannelSegment(
  object: ChannelSegment<TimeSeries>
): object is ChannelSegment<Waveform> {
  return object.timeseriesType === TimeSeriesType.WAVEFORM;
}

/**
 * Checks if waveform channel segment OSD representation
 * @param object Channel segment OSD representation
 * @returns boolean
 */
export function isOSDWaveformChannelSegment(
  object: OSDChannelSegment<OSDTimeSeries>
): object is OSDChannelSegment<OSDWaveform> {
  return object.timeseriesType === TimeSeriesType.WAVEFORM;
}

/**
 * Checks if channel segment OSD representation
 * @param object Channel segment or channel segment OSD representation
 * @returns boolean
 */
export function isOSDChannelSegment(
  object: ChannelSegment<TimeSeries> | OSDChannelSegment<OSDTimeSeries>
): object is OSDChannelSegment<OSDTimeSeries> {
  return typeof object.startTime === 'string';
}

/**
 * Checks if channel segment
 * @param object Channel segment or channel segment OSD representation
 * @returns boolean
 */
export function isChannelSegment(
  object: ChannelSegment<TimeSeries> | OSDChannelSegment<OSDTimeSeries>
): object is ChannelSegment<TimeSeries> {
  return typeof object.startTime === 'number';
}

/**
 * Checks if time series OSD representation
 * @param object Time series or Time Series OSD representation
 */
export function isOSDTimeSeries(object: TimeSeries | OSDTimeSeries): object is OSDTimeSeries {
  return typeof object.startTime === 'string';
}

/**
 * Checks if time series
 * @param object Time series or time series OSD representation
 * @returns boolean
 */
export function isTimeSeries(object: TimeSeries | OSDTimeSeries): object is TimeSeries {
  return typeof object.startTime === 'number';
}
