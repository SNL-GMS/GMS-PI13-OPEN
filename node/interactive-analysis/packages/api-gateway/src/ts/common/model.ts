import { LogLevel } from '@gms/common-graphql/lib/graphql/common/types';

/**
 * Common model definitions shared across gateway data APIs
 */

/**
 * Represents an association mapping
 * between an event and signal detection.
 */
export interface Association {
  /** the event association id */
  readonly associationId: string;
  /** the event id */
  readonly eventId: string;
  /** the event hypothesis id */
  readonly eventHypothesisId: string;
  /** the signal detection id */
  readonly signalDetectionId: string;
  /** the signal detection hypothesis id */
  readonly signalDetectionHypothesisId: string;
  /** the association reject flag */
  readonly rejected: boolean;
}

/**
 * Represents a location specified using latitude (degrees), longitude (degrees),
 * and altitude (kilometers).
 */
export interface Location {
  readonly latitudeDegrees: number;
  readonly longitudeDegrees: number;
  readonly elevationKm: number;
  readonly depthKm?: number;
}

/**
 * Represents a frequency range
 */
export interface FrequencyBand {
  readonly minFrequencyHz: number;
  readonly maxFrequencyHz: number;
}

/**
 * Time range
 */
export interface TimeRange {
  readonly startTime: number;
  readonly endTime: number;
}

/**
 * Keeps track of state between different workspaces
 */
export interface WorkspaceState {
  readonly eventToUsers: EventToUsers[];
}

/**
 * Event with list of usernames
 */
export interface EventToUsers {
  readonly eventId: string;
  readonly userNames: string[];
}

/**
 * Enumeration representing the different types of stations in the monitoring network.
 */
export enum StationType {
  Seismic3Component = 'Seismic3Component',
  Seismic1Component = 'Seismic1Component',
  SeismicArray = 'SeismicArray',
  Hydroacoustic = 'Hydroacoustic',
  HydroacousticArray = 'HydroacousticArray',
  Infrasound = 'Infrasound',
  InfrasoundArray = 'InfrasoundArray',
  Weather = 'Weather',
  UNKNOWN = 'UNKNOWN'
}

/**
 * Enumeration representing the different types of processing channels.
 */
export enum ChannelDataType {
  SEISMIC = 'SEISMIC',
  HYDROACOUSTIC = 'HYDROACOUSTIC',
  INFRASOUND = 'INFRASOUND',
  WEATHER = 'WEATHER',
  DIAGNOSTIC_SOH = 'DIAGNOSTIC_SOH',
  DIAGNOSTIC_WEATHER = 'DIAGNOSTIC_WEATHER'
}

/**
 * Processing Context part of data structure that filtered waveform
 * control service uses to tell who/where the request came from. In our case
 * from Interactive UI user
 */
export interface ProcessingContext {
  readonly analystActionReference: AnalystActionReference;
  readonly processingStepReference: ProcessingStepReference;
  readonly storageVisibility: string;
}

/**
 * Analyst action reference
 */
export interface AnalystActionReference {
  readonly processingStageIntervalId: string;
  readonly processingActivityIntervalId: string;
  readonly analystId: string;
}

/**
 * Processing step reference
 */
export interface ProcessingStepReference {
  readonly processingStageIntervalId: string;
  readonly processingSequenceIntervalId: string;
  readonly processingStepId: string;
}

/**
 * Test data paths used when reading in data
 */
export interface TestDataPaths {
  readonly dataHome: string;
  readonly jsonHome: string;
  readonly fpHome: string;
  readonly fkHome: string;
  readonly channelsHome: string;
  readonly additionalDataHome: string;
  readonly integrationDataHome: string;
  readonly tempTigerTeamData: string;
}

/**
 * Represents calibration information associated with a waveform
 */
export interface ProcessingCalibration {
  readonly factor: number;
  readonly factorError: number;
  readonly period: number;
  readonly timeShift: number;
}

/**
 * Relative Position information relative to a location
 */
export interface Position {
  readonly northDisplacementKm: number;
  readonly eastDisplacementKm: number;
  readonly verticalDisplacementKm: number;
}

/**
 * Represents the configured type of data source the API Gateway provides access to - values:
 * Local - The API gateway loads data from local file storage for testing purposes
 * Service - The API gateway uses services to provide access to backend (e.g. OSD) data
 */
export enum AccessorDataSource {
  Local = 'Local',
  Service = 'Service'
}

/**
 * Enumerated list of source types used to compute distances to
 */
export enum DistanceSourceType {
  Event = 'Event',
  UserDefined = 'UserDefined'
}

/**
 * Distance value's units degrees or kilometers
 */
export enum DistanceUnits {
  degrees = 'degrees',
  km = 'km'
}

/**
 * The distance value representing degrees and km.
 */
export interface Distance {
  readonly degrees: number;
  readonly km: number;
}

/**
 * Represents input arguments for calculating distance measurement
 * relative to a specified source location
 */
export interface DistanceToSourceInput {
  // The type of the source the distance is measured to (e.g. and event)
  readonly sourceType: DistanceSourceType;

  // the unique ID of the source object
  readonly sourceId: string;
}

/**
 * Represents a distance measurement relative to a specified source location
 */
export interface DistanceToSource {
  // The distance
  readonly distance: Distance;

  // The azimuth
  readonly azimuth: number;

  // The source location
  readonly sourceLocation: Location;

  // The type of the source the distance is measured to (e.g. and event)
  readonly sourceType: DistanceSourceType;

  // the unique ID of the source object
  readonly sourceId: string;

  // Which station distance to the source
  readonly stationId: string;
}

/**
 * Client log, object that describes the client log message
 */
export interface ClientLog {
  readonly logLevel: LogLevel;
  readonly message: string;
  readonly time?: string;
}

/**
 * Units used in DoubleValue part of feature prediction
 */
export enum Units {
  DEGREES = 'DEGREES',
  RADIANS = 'RADIANS',
  SECONDS = 'SECONDS',
  HERTZ = 'HERTZ',
  SECONDS_PER_DEGREE = 'SECONDS_PER_DEGREE',
  SECONDS_PER_RADIAN = 'SECONDS_PER_RADIAN',
  SECONDS_PER_DEGREE_SQUARED = 'SECONDS_PER_DEGREE_SQUARED',
  SECONDS_PER_KILOMETER_SQUARED = 'SECONDS_PER_KILOMETER_SQUARED',
  SECONDS_PER_KILOMETER = 'SECONDS_PER_KILOMETER',
  SECONDS_PER_KILOMETER_PER_DEGREE = 'SECONDS_PER_KILOMETER_PER_DEGREE',
  ONE_OVER_KM = 'ONE_OVER_KM',
  NANOMETERS = 'NANOMETERS',
  NANOMETERS_PER_SECOND = 'NANOMETERS_PER_SECOND',
  NANOMETERS_PER_COUNT = 'NANOMETERS_PER_COUNT',
  UNITLESS = 'UNITLESS',
  MAGNITUDE = 'MAGNITUDE',
  COUNTS_PER_NANOMETER = 'COUNTS_PER_NANOMETER',
  COUNTS_PER_PASCAL = 'COUNTS_PER_PASCAL',
  PASCALS_PER_COUNT = 'PASCALS_PER_COUNT'
}

/**
 * Phase type list
 */
export enum PhaseType {
  // TODO: need to elaborate with full set of phase labels
  P = 'P',
  S = 'S',
  P3KPbc = 'P3KPbc',
  P4KPdf_B = 'P4KPdf_B',
  P7KPbc = 'P7KPbc',
  P7KPdf_D = 'P7KPdf_D',
  PKiKP = 'PKiKP',
  PKKSab = 'PKKSab',
  PKP2bc = 'PKP2bc',
  PKP3df_B = 'PKP3df_B',
  PKSab = 'PKSab',
  PP_1 = 'PP_1',
  pPKPbc = 'pPKPbc',
  PS = 'PS',
  Rg = 'Rg',
  SKiKP = 'SKiKP',
  SKKSac = 'SKKSac',
  SKPdf = 'SKPdf',
  SKSdf = 'SKSdf',
  sPdiff = 'sPdiff',
  SS = 'SS',
  sSKSdf = 'sSKSdf',
  Lg = 'Lg',
  P3KPbc_B = 'P3KPbc_B',
  P5KPbc = 'P5KPbc',
  P7KPbc_B = 'P7KPbc_B',
  Pb = 'Pb',
  PKKP = 'PKKP',
  PKKSbc = 'PKKSbc',
  PKP2df = 'PKP2df',
  PKPab = 'PKPab',
  PKSbc = 'PKSbc',
  PP_B = 'PP_B',
  pPKPdf = 'pPKPdf',
  PS_1 = 'PS_1',
  SKKP = 'SKKP',
  SKKSac_B = 'SKKSac_B',
  SKS = 'SKS',
  SKSSKS = 'SKSSKS',
  sPKiKP = 'sPKiKP',
  SS_1 = 'SS_1',
  SSS = 'SSS',
  nNL = 'nNL',
  P3KPdf = 'P3KPdf',
  P5KPbc_B = 'P5KPbc_B',
  P7KPbc_C = 'P7KPbc_C',
  PcP = 'PcP',
  PKKPab = 'PKKPab',
  PKKSdf = 'PKKSdf',
  PKP3 = 'PKP3',
  PKPbc = 'PKPbc',
  PKSdf = 'PKSdf',
  pPdiff = 'pPdiff',
  PPP = 'PPP',
  pSdiff = 'pSdiff',
  Sb = 'Sb',
  SKKPab = 'SKKPab',
  SKKSdf = 'SKKSdf',
  SKS2 = 'SKS2',
  Sn = 'Sn',
  sPKP = 'sPKP',
  SS_B = 'SS_B',
  SSS_B = 'SSS_B',
  NP = 'NP',
  P3KPdf_B = 'P3KPdf_B',
  P5KPdf = 'P5KPdf',
  P7KPdf = 'P7KPdf',
  PcS = 'PcS',
  PKKPbc = 'PKKPbc',
  PKP = 'PKP',
  PKP3ab = 'PKP3ab',
  PKPdf = 'PKPdf',
  Pn = 'Pn',
  pPKiKP = 'pPKiKP',
  PPP_B = 'PPP_B',
  pSKS = 'pSKS',
  ScP = 'ScP',
  SKKPbc = 'SKKPbc',
  SKP = 'SKP',
  SKS2ac = 'SKS2ac',
  SnSn = 'SnSn',
  sPKPab = 'sPKPab',
  sSdiff = 'sSdiff',
  NP_1 = 'NP_1',
  P4KPbc = 'P4KPbc',
  P5KPdf_B = 'P5KPdf_B',
  P7KPdf_B = 'P7KPdf_B',
  Pdiff = 'Pdiff',
  PKKPdf = 'PKKPdf',
  PKP2 = 'PKP2',
  PKP3bc = 'PKP3bc',
  PKPPKP = 'PKPPKP',
  PnPn = 'PnPn',
  pPKP = 'pPKP',
  PPS = 'PPS',
  pSKSac = 'pSKSac',
  ScS = 'ScS',
  SKKPdf = 'SKKPdf',
  SKPab = 'SKPab',
  SKS2df = 'SKS2df',
  SP = 'SP',
  sPKPbc = 'sPKPbc',
  sSKS = 'sSKS',
  P4KPdf = 'P4KPdf',
  P5KPdf_C = 'P5KPdf_C',
  P7KPdf_C = 'P7KPdf_C',
  Pg = 'Pg',
  PKKS = 'PKKS',
  PKP2ab = 'PKP2ab',
  PKP3df = 'PKP3df',
  PKS = 'PKS',
  PP = 'PP',
  pPKPab = 'pPKPab',
  PPS_B = 'PPS_B',
  pSKSdf = 'pSKSdf',
  Sdiff = 'Sdiff',
  SKKS = 'SKKS',
  SKPbc = 'SKPbc',
  SKSac = 'SKSac',
  SP_1 = 'SP_1',
  sPKPdf = 'sPKPdf',
  sSKSac = 'sSKSac',
  Sx = 'Sx',
  tx = 'tx',
  N = 'N',
  Px = 'Px',
  PKhKP = 'PKhKP',
  UNKNOWN = 'UNKNOWN',
  LR = 'LR'
}
