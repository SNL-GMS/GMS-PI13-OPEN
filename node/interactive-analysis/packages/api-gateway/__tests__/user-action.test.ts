import { UserActionDescription } from '../src/ts/cache/model';
import {
  userActionCreatorForEventHypothesisChange,
  userActionCreatorForSignalDetectionHypothesisChange
} from '../src/ts/cache/user-action';
import { FeatureMeasurementTypeName } from '../src/ts/signal-detection/model';

// Test data for this test is defined below the tests

// Mostly empty signal detection for testing user actions
const dummyDetection: any = {
  stationName: 'TEST'
};

// Mostly empty event hypothesis for testing user actions
const newEventHypothesis: any = {
  preferredLocationSolution: {
    locationSolution: {
      location: {
        time: 1274349600
      }
    }
  }
};

// Mostly empty signal detection hypothesis for testing user actions
const oldSDHypothesis: any = {
  featureMeasurements: [
    {
      featureMeasurementType: FeatureMeasurementTypeName.PHASE,
      measurementValue: {
        phase: 'P'
      }
    },
    {
      featureMeasurementType: FeatureMeasurementTypeName.ARRIVAL_TIME,
      measurementValue: {
        value: 1274349600
      }
    },
    {
      featureMeasurementType: FeatureMeasurementTypeName.AMPLITUDE_A5_OVER_2,
      measurementValue: {
        amplitude: {
          value: 1.23
        }
      }
    }
  ]
};

// Mostly empty signal detection hypothesis for testing user actions
const newSDHypothesis: any = {
  featureMeasurements: [
    {
      featureMeasurementType: FeatureMeasurementTypeName.PHASE,
      measurementValue: {
        phase: 'S'
      }
    },
    {
      featureMeasurementType: FeatureMeasurementTypeName.ARRIVAL_TIME,
      measurementValue: {
        value: 12743496720
      }
    },
    {
      featureMeasurementType: FeatureMeasurementTypeName.AMPLITUDE_A5_OVER_2,
      measurementValue: {
        amplitude: {
          value: 0.23
        }
      }
    }
  ]
};

describe('User Action Tests', () => {
  it('Snapshot for signal detection actions', () => {
    Object.values(UserActionDescription).forEach(action => {
      const userAction = userActionCreatorForSignalDetectionHypothesisChange(
        action,
        dummyDetection,
        oldSDHypothesis,
        newSDHypothesis
      );
      expect(userAction.toString()).toMatchSnapshot();
    });
  });

  it('Snapshot for event actions', () => {
    Object.values(UserActionDescription).forEach(action => {
      const userAction = userActionCreatorForEventHypothesisChange(
        action,
        undefined,
        newEventHypothesis
      );
      expect(userAction.toString()).toMatchSnapshot();
    });
  });
});
