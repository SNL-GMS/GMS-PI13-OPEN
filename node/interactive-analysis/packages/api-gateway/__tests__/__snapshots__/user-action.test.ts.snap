// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Action Tests Snapshot for event actions 1`] = `"Unknown"`;

exports[`User Action Tests Snapshot for event actions 2`] = `"Update event from signal detection change"`;

exports[`User Action Tests Snapshot for event actions 3`] = `"Update event from signal detection change"`;

exports[`User Action Tests Snapshot for event actions 4`] = `"Update event from signal detection change"`;

exports[`User Action Tests Snapshot for event actions 5`] = `"Update event from signal detection change"`;

exports[`User Action Tests Snapshot for event actions 6`] = `"Update event from signal detection change"`;

exports[`User Action Tests Snapshot for event actions 7`] = `"Update event from signal detection change"`;

exports[`User Action Tests Snapshot for event actions 8`] = `"Update event from signal detection change"`;

exports[`User Action Tests Snapshot for event actions 9`] = `"Update event from signal detection change"`;

exports[`User Action Tests Snapshot for event actions 10`] = `"Update event from signal detection change"`;

exports[`User Action Tests Snapshot for event actions 11`] = `"Update event from signal detection change"`;

exports[`User Action Tests Snapshot for event actions 12`] = `"Update event from signal detection change"`;

exports[`User Action Tests Snapshot for event actions 13`] = `"Update event from signal detection change"`;

exports[`User Action Tests Snapshot for event actions 14`] = `"Update event from signal detection change"`;

exports[`User Action Tests Snapshot for event actions 15`] = `"Update event from signal detection change"`;

exports[`User Action Tests Snapshot for event actions 16`] = `"Update event from signal detection change"`;

exports[`User Action Tests Snapshot for event actions 17`] = `"Create event at 10:00:00"`;

exports[`User Action Tests Snapshot for event actions 18`] = `"Locate event at 10:00:00"`;

exports[`User Action Tests Snapshot for event actions 19`] = `"Update event from signal detection change"`;

exports[`User Action Tests Snapshot for event actions 20`] = `"Update event feature predictions"`;

exports[`User Action Tests Snapshot for event actions 21`] = `"Event opened for refinement"`;

exports[`User Action Tests Snapshot for event actions 22`] = `"Event marked complete"`;

exports[`User Action Tests Snapshot for event actions 23`] = `"Update event preferred hypothesis"`;

exports[`User Action Tests Snapshot for event actions 24`] = `"Change Magnitude defining settings"`;

exports[`User Action Tests Snapshot for event actions 25`] = `"Save event"`;

exports[`User Action Tests Snapshot for signal detection actions 1`] = `"Unknown"`;

exports[`User Action Tests Snapshot for signal detection actions 2`] = `"Associate S on TEST"`;

exports[`User Action Tests Snapshot for signal detection actions 3`] = `"Associate S on TEST"`;

exports[`User Action Tests Snapshot for signal detection actions 4`] = `"Unassociate S on TEST"`;

exports[`User Action Tests Snapshot for signal detection actions 5`] = `"Unassociate S on TEST"`;

exports[`User Action Tests Snapshot for signal detection actions 6`] = `"Create detection S on TEST at 04:12:00"`;

exports[`User Action Tests Snapshot for signal detection actions 7`] = `"Reject detection S on TEST"`;

exports[`User Action Tests Snapshot for signal detection actions 8`] = `"Reject detection S on TEST"`;

exports[`User Action Tests Snapshot for signal detection actions 9`] = `"Time S from 10:00:00 to 04:12:00 on TEST"`;

exports[`User Action Tests Snapshot for signal detection actions 10`] = `"Phase from P to S on TEST"`;

exports[`User Action Tests Snapshot for signal detection actions 11`] = `"Phase from P to S on TEST"`;

exports[`User Action Tests Snapshot for signal detection actions 12`] = `"Amplitude for S on TEST from 1.230 to 0.230"`;

exports[`User Action Tests Snapshot for signal detection actions 13`] = `"Update detection review amplitude"`;

exports[`User Action Tests Snapshot for signal detection actions 14`] = `"Update detection"`;

exports[`User Action Tests Snapshot for signal detection actions 15`] = `"Fk S on TEST"`;

exports[`User Action Tests Snapshot for signal detection actions 16`] = `"Fk S on TEST"`;

exports[`User Action Tests Snapshot for signal detection actions 17`] = `"Associate S on TEST"`;

exports[`User Action Tests Snapshot for signal detection actions 18`] = `"Locate event"`;

exports[`User Action Tests Snapshot for signal detection actions 19`] = `"Update event from signal detection change"`;

exports[`User Action Tests Snapshot for signal detection actions 20`] = `"Update event feature predictions"`;

exports[`User Action Tests Snapshot for signal detection actions 21`] = `"Event opened for refinement"`;

exports[`User Action Tests Snapshot for signal detection actions 22`] = `"Event marked complete"`;

exports[`User Action Tests Snapshot for signal detection actions 23`] = `"Update event preferred hypothesis"`;

exports[`User Action Tests Snapshot for signal detection actions 24`] = `"Change Magnitude defining settings"`;

exports[`User Action Tests Snapshot for signal detection actions 25`] = `"Save event"`;
